### PR TITLE
Compile.rst: major updates to Windows and OSX sections;  

### PR DESCRIPTION
**Pull request by TheBloke**
minor to Linux; new section re compiling documentation; many small fixes.

Introduction.rst: fix typo and remove erroneous end version for OSX.

All changes have been tested as compiled HTML.

### CODE REVIEW

File: docs/Compile.rst [hunk 1]
```diff
@@ -25,9 +26,26 @@ To get the code::
...
+To get the latest development code (develop branch)::
+
+  git clone --recursive https://github.com/DFHack/dfhack
+  cd dfhack
+  git checkout develop
+  git submodule update
+
+**Important note on submodule update**:
+
+It is necessary to run ``git submodule update`` every time you change Git branch,
```
Comment (lethosor) [inline docs/Compile.rst]: Grammar?

File: docs/Compile.rst [hunk 2]
```diff
@@ -240,12 +392,123 @@ Next, run one of the scripts with ``generate`` prefix. These create the MSVC sol
...
+In order to build the documentation, you must have Python with Sphinx
+version 1.3.1 or later. Both Python 2.x and 3.x are supported.
+
+When installing Sphinx from OS package managers, be aware that there is
+another program called Sphinx, completely unrelated to documentation management.
+Be sure you are installing the right Sphinx!
+
+
+Linux
+-----PTION(BUILD_DOCS "Choose whether to build the documentation (requires python and Sphinx)." ON)
```
Comment (lethosor) [inline docs/Compile.rst]: s/-----/O/

File: docs/Compile.rst [hunk 3]
```diff
@@ -240,12 +392,123 @@ Next, run one of the scripts with ``generate`` prefix. These create the MSVC sol
...
+
+
+Required dependencies
+=====================
+In order to build the documentation, you must have Python with Sphinx
+version 1.3.1 or later. Both Python 2.x and 3.x are supported.
+
+When installing Sphinx from OS package managers, be aware that there is
+another program called Sphinx, completely unrelated to documentation management.
+Be sure you are installing the right Sphinx!
```
Comment (lethosor) [inline docs/Compile.rst]: Often python-sphinx, I think

File: docs/Compile.rst [hunk 4]
```diff
@@ -25,9 +26,26 @@ To get the code::
 If your version of git does not support the ``--recursive`` flag, you will need to omit it and run
 ``git submodule update --init`` after entering the dfhack directory.
 
+To get the latest development code (develop branch)::
+
+  git clone --recursive https://github.com/DFHack/dfhack
+  cd dfhack
```
Comment (lethosor) [inline docs/Compile.rst]: I'm not entirely sure if repeating these first two lines is necessary.

**Comment by TheBloke** [comment]
OK check again

**Comment by TheBloke** [comment]
Or rather do now that I've actually committed the fixes

**Comment by TheBloke** [comment]
Few more changes

File: docs/Compile.rst [hunk 5]
```diff
@@ -10,14 +10,15 @@ and `install the latest release instead <installing>`.
 .. contents::
    :depth: 2
 
+.. _compile-how-to-get-the-code:
 
 How to get the code
 ===================
 DFHack doesn't have any kind of system of code snapshots in place, so you will have to
-get code from the github repository using git.  How to get git is described under
+get code from the Github repository using git.  How to get git is described under
```
Comment (dscorbett) [inline docs/Compile.rst]: `s/Github/GitHub/; s/git/Git/g`

Comment (TheBloke) [inline docs/Compile.rst]: Thanks, changed.

**Comment by expwnent** [comment]
Are you sure Git is supposed to be capitalized when not a command? You don't capitalize ls, by comparison.

**Comment by dscorbett** [comment]
Yes, Git is supposed to be capitalized. See http://git-scm.com/.

**Comment by TheBloke** [comment]
I'd originally thought not, but after dscrobett's comment I Googled and Git as a product/package is capitalised.

**Comment by TheBloke** [comment]
OK I'm definitely done now, closing my Vim window so I stop tweaking.  Let me know if there's any more issues or desirable changes.

File: docs/Compile.rst [hunk 6]
```diff
@@ -51,15 +71,19 @@ Dependencies
...
-work due to ABI changes (the entire plugin loading system won't work, for
-example). On 64-bit distributions, you'll need the multilib development tools
-and libraries (``gcc-multilib`` or ``gcc-4.x-multilib`` on Debian). Note that
-installing a 32-bit GCC on 64-bit systems (e.g. ``gcc:i386`` on Debian) will
+(see below), but any version from 4.5 onwards (including 5.x) will work.
+
+On 64-bit distributions, you'll need the multilib development tools and libraries:
+
+* ``gcc-multilib`` and ``g++-multilib``
+* On Debian: ``gcc-4.x-multilib`` and ``g++-4.x-multilib``
```
Comment (lethosor) [inline docs/Compile.rst]: No, it's still `gcc-multilib` and `g++-multilib` (I'm pretty sure gcc-4.x isn't a package). The 4.x part is necessary if you're using GCC 4.x (e.g. use gcc-4.9-multilib and g++-4.9-multilib if you're compiling with GCC 4.9).

File: docs/Compile.rst [hunk 7]
```diff
@@ -121,7 +156,9 @@ DF, which causes DF to use your system libstdc++ instead::
 Note that distributing binaries compiled with newer GCC versions requires end-
 users to delete libstdc++ themselves and have a libstdc++ on their system from
 the same GCC version or newer. For this reason, distributing anything compiled
-with GCC versions newer than 4.8 is discouraged.
```
Comment (lethosor) [inline docs/Compile.rst]: The first version we had issues with was 4.9, since Arch adopted it before Debian, so the 4.8 was intentional because it should work on pretty much any reasonably up-to-date system. However, if you have to install a different version of gcc for compatibility, an older version (>=4.5) would probably be better, although support for 4.5-4.7 (maybe 4.8 too) ought to be equally widely available on different distros.

File: docs/Compile.rst [hunk 8]
```diff
@@ -134,12 +171,16 @@ following environment variable::
 
     export MACOSX_DEPLOYMENT_TARGET=10.9
 
-#. Download and unpack a copy of the latest DF
-#. Install Xcode from Mac App Store
-#. Open Xcode, go to Preferences > Downloads, and install the Command Line Tools.
-#. Install dependencies
+* Download and unpack a copy of the latest DF
+* Install Xcode from Mac App Store
```
Comment (lethosor) [inline docs/Compile.rst]: Might want to mention 10.11 issues sooner (here or above?)

File: docs/Compile.rst [hunk 9]
```diff
@@ -200,39 +280,141 @@ On Windows, DFHack replaces the SDL library distributed with DF.
...
+and check for and install the SP1 update. If no update is found, check that
+your Windows Update settings include "Updates from all Microsoft products".
+
+You can confirm whether you have SP1 by opening the Visual Studio 2010 IDE
+and selecting About from the Help menu.  If you have SP1 it will have *SP1Rel*
+at the end of the version number, for example: *Version 10.0.40219.1 SP1Rel*
+
+It is vital that you use SP1 as, while building with the original release
+of Visual Studio 2010 (RTM) may succeed, it will result in non-working DFHack
+binaries that crash when connecting to Dwarf Fortress.
```
Comment (lethosor) [inline docs/Compile.rst]: I thought missing SP1 just led to compilation issues but the ABIs were compatible. Did SP1 also break ABI compatibility?

("interacting with" might be better than "connecting to", since DFHack runs in-process.)

File: docs/Compile.rst [hunk 10]
```diff
@@ -200,39 +280,141 @@ On Windows, DFHack replaces the SDL library distributed with DF.
...
+
+Additional dependencies: installing with the Chocolatey Package Manager
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The remainder of dependencies - Git, CMake and StrawberryPerl - can be most
+easily installed using the Chocolatey Package Manger. Chocolatey is a
+\*nix-style package manager for Windows. Think "``apt-get`` for Windows."
+
+Chocolatey is the recommended way of installing the required dependencies
```
Comment (lethosor) [inline docs/Compile.rst]: I'm a bit hesitant to refer to this as "the" recommended way, unless there was a discussion that led to that which I'm not aware of.

File: docs/Compile.rst (hunk 6 above)
Comment (TheBloke) [inline docs/Compile.rst]: OK, so just remove those Debian lines?  That's not anything I've edited, just brought over from the original version of the doc (with updated formatting.)

File: docs/Compile.rst (hunk 9 above)
Comment (TheBloke) [inline docs/Compile.rst]: Ah, I thought it was ABI issues on the basis that it *does* compile on RTM (non-SP1).   At least someone in #dfhack got it compiled the other day on RTM.  He did have a bunch of issues but ultimately a lot of that may have been his own system issues.

If you've heard it's not ABI issues, and I've heard it's not compilation issues, then maybe it's no issue at all.

I can try it out I guess, but it's so easy for anyone to upgrade to SP1 I'll probably just tone down the above message to something like "Using SP1 may cause issues, don't do it." rather than stating what they are.

File: docs/Compile.rst (hunk 10 above)
Comment (TheBloke) [inline docs/Compile.rst]: No discussion has occurred, it's merely my own strong recommendation.

Chocolatey installs the exact same files as one would manually, just it sets up PATHs etc correctly, installs any needed dependencies, provides ongoing application management (upgrades etc), and is a useful system to have in place for any more complex requirements eg when I reference it under Docs as a one-line install for Python/Pip for Sphinx.

Chocolatey has been around 4.5 years and has 3000+ unique packages with a pretty active community, so it's pretty stable.  And we've had known issues in #dfhack with people not getting PATHs set up right, so known-good config is a strong benefit in my view.  So I'd have to say it's little different to recommending apt-get as the standard for Ubuntu. 

But I can tone down the nature of the recommendation if you think it's necessary.

File: docs/Compile.rst (hunk 7 above)
Comment (TheBloke) [inline docs/Compile.rst]: OK I'm a bit confused about what needs to be said here.

My understanding was that:
* If we're distributing, then currently only 4.5 should be used until such time as we look at libstdc++ distribution
* If it's for personal, anything >= 4.5 is fine, right up to current 5.2.1 and, according to GCC's stated ABI compat, beyond for the foreseeable future
  *  and there's a strong benefit to using a later version on modern distros, simply in that it's installed by default and one doesn't need to spend the time working out how to install an older version, mess with the alternatives system, etc.

I'm not sure where a 4.7 or 4.8 or anything-other-than-4.5 fits in?  If you have to install an old version of GCC, surely it's only for distribution purposes and therefore must be 4.5?  And if you're not distributing, you can use any later version and it's easier for you if you do so.  

So if a user happens to be on a 4.7, 4.8 or 4.9 system then that's fine, just as it is if they're on 5.x, but there's no reason anyone would want to install 4.x unless it's 4.5 for distributing?  So is there a need to mention any version besides 4.5 still?

(As an aside, it just occurred to me that it might be useful to have a new "Note on using GCC 5.x" section which explicitly mentions the ABI compatibility flag that cmake is applying.  This relates to potential new contributors, who must be aware that they need to compile their own plugins with that flag if they're testing on GCC 5 and plan to work with DFHack.  That might belong in Contributing.rst, instead of or as well as this Compile.rst,)

File: docs/Compile.rst (hunk 9 above)
Comment (TheBloke) [inline docs/Compile.rst]: I've made it vague and general in the next commit, merely stating that SP1 "has been reported to cause issues" and is "not supported by DFHack".

File: docs/Compile.rst (hunk 8 above)
Comment (TheBloke) [inline docs/Compile.rst]: OK done in next commit

File: docs/Compile.rst (hunk 6 above)
Comment (TheBloke) [inline docs/Compile.rst]: Just removed the extra "On Debian" line in next commit

File: docs/Compile.rst (hunk 7 above)
Comment (lethosor) [inline docs/Compile.rst]: Sorry, I was primarily talking about distribution there (on Linux only). Pretty much any linux distro at this point should default to GCC 4.8 or newer, so compiling with GCC 4.5-4.8 should result in binaries that work on a majority of distros. We ran into issues in the past where Arch moved to GCC 4.9 but Debian stayed at 4.8, for example, and in that case GCC 4.9 wasn't even available in apt, so a couple DFHack builds broke completely on a lot of distros.
In short:
* GCC 4.5 produces builds that should work for everyone, but probably needs to be compiled from source or obtained from a third-party package repo
* GCC 4.6-4.8 (and maybe 4.9 by now?) produce builds that work for nearly everyone (with a new enough libstdc++), but users have to get rid of DF's libstdc++
I haven't investigated the availability of GCC 5 yet, but I'd say sticking with 4.5-4.8 (or 4.9) is best for releases for now.

File: docs/Compile.rst (hunk 10 above)
Comment (lethosor) [inline docs/Compile.rst]: I don't know how much time/space it takes to install, though - we've also had people complain about the number/size of dependencies on Windows, so I don't want to make it sound like additional ones are required when they technically aren't. (It sounds useful, though, so do encourage it.)

File: docs/Compile.rst (hunk 7 above)
Comment (lethosor) [inline docs/Compile.rst]: (Of course, the system GCC is fine for personal builds as long as you get rid of DF's libstdc++.)

Comment (TheBloke) [inline docs/Compile.rst]: OK great.  I'm pretty sure what I have covers that already, but let me know if it isn't clear or needs more explanation.

Or rather, what I have basically says "for now only 4.5 for distribution, anything >=4.5 for personal"

So I could clarify that actually 4.6, 4.7 and 4.8 might be OK for distribution, but I'm not sure that's necessary given distribution only needs to be available for a few people, all of whom have 4.5 available anyway.

File: docs/Compile.rst (hunk 10 above)
Comment (TheBloke) [inline docs/Compile.rst]: Chocolatey itself is using 18.5MB of disk space on my system with 20 or so packages installed.

I'd guess it takes maybe five minutes to install including opening/closing the cmd.exe windows etc.  Definitely significantly less time than it takes to install the three individual dependencies, factoring in time to go to three URLs, find the download link, download them, double click on them, click Next through the various installer wizards, then manually add them to PATH where not done already (Perl definitely needs manually adding, CMake doesn't, not sure about Git.)   

I'd say it's maybe 5 minutes total time for the Chocolatey + all dependencies install, versus at least 10 for the manual method, maybe 15.

I'd say it has about the same impact in terms of space and time as Homebrew or MacPorts on OSX - really not any at all.

But I'll reword it to highlight it's not actually required. Maybe 'preferred' instead of recommended.

Comment (TheBloke) [inline docs/Compile.rst]: OK I was curious so I timed it on a separate PC.

3:43 to install Chocolatey + and three dependencies.  Of that, 45 seconds was the time to install Chocolatey itself, and about two minutes was the time it took to download the 75MB StrawberryPerl package.

Basically it's just as fast and easy as you'd expect any binary-based package manager to be.

Comment (TheBloke) [inline docs/Compile.rst]: Incidentally, if we're ever able to move to MSVC >2010, there are Chocolatey packages to silently install the free versions of Visual Studio versions 2012->2015.

I could probably make and submit a Chocolatey formula to do that for 2010 as well, however the version we found was distributed as an ISO and I'm not sure if Windows 7 can auto-mount that (I haven't yet checked to see if Chocolatey has other ISO-based formula and therefore if they have a way to deal with that.)

But definitely if we ever get off 2010, Windows can have all its dependencies handled in a single command line same as Linux and OSX.

Comment (TheBloke) [inline docs/Compile.rst]: OK updated it in the next commit, now it is "a preferred way of installing the required dependencies".  And I put in a brief note on the very small disk requirements (I checked on the Windows 7 install in which I installed Chocolatey + just the 3 DFHack dependencies, and disk usage is 7.5MB there.)  And similarly removed the strong recommendation for Chocolatey under the Manually Installing Dependencies section.
